cli/src/cmd_disk.rs

```diff
@@ -93,461 +78,107 @@ pub struct CmdDiskImport {
+        if self.snapshot.is_some() {
+            // Clap enforces that all of these are present when snapshot is.
+            builder = builder.image_info(ImageInfo {
+                snapshot: self.snapshot.clone().unwrap(),
+                image: self.image.clone().unwrap(),
+                image_description: self.image_description.clone().unwrap(),
+                image_os: self.image_os.clone().unwrap(),
+                image_version: self.image_version.clone().unwrap(),
+            });
         }
 
-        // Create one tokio task for each thread that will upload file chunks
-        let mut handles: Vec<tokio::task::JoinHandle<Result<()>>> =
-            Vec::with_capacity(UPLOAD_TASKS);
-        let mut senders = Vec::with_capacity(UPLOAD_TASKS);
-
-        let pb = Arc::new(ProgressBar::new(file_size));
-        pb.set_style(ProgressStyle::default_bar()
-            .template("[{elapsed_precise}] [{wide_bar:.green}] {bytes}/{total_bytes} ({bytes_per_sec}, {eta})")?
-        );
-        pb.set_position(0);
-
-        for _i in 0..UPLOAD_TASKS {
-            let (tx, mut rx) = mpsc::channel(100);
-
-            let client = client.clone();
-            let disk_name = self.disk.clone();
-            let project = self.project.clone();
-
-            let pb = pb.clone();
-
-            handles.push(tokio::spawn(async move {
-                while let Some((offset, base64_encoded_data, data_len)) = rx.recv().await {
-                    client
-                        .disk_bulk_write_import()
-                        .disk(disk_name.clone())
-                        .project(project.clone())
-                        .body(ImportBlocksBulkWrite {
-                            offset,
-                            base64_encoded_data,
-                        })
-                        .send()
-                        .await?;
-
-                    pb.inc(data_len as u64);
-                }
+        let (import_future, handle) = builder.execute()?;
 
-                Ok(())
-            }));
+        let pb = start_progress_bar(handle.progress(), disk_info.file_size, &self.disk)?;
+        watch_for_ctrl_c(handle, pb);
 
-            senders.push(tx);
-        }
-
-        // Read chunks from the file in the file system and send them to the
-        // upload threads.
-        let mut file = File::open(&self.path)?;
-        let mut i = 0;
-        let mut offset = 0;
-
-        let read_result: Result<()> = loop {
-            let mut chunk = Vec::with_capacity(CHUNK_SIZE as usize);
-
-            let n = match file.by_ref().take(CHUNK_SIZE).read_to_end(&mut chunk) {
-                Ok(n) => n,
-                Err(e) => {
-                    eprintln_nopipe!(
-                        "reading from {} failed with {:?}",
-                        self.path.to_string_lossy(),
-                        e,
-                    );
-                    break Err(e.into());
-                }
-            };
+        import_future.await?;
```

Comment:
I'm not sure if this is going to be more complex or simpler, but what do you think of doing something like this:

```
let real_work_handle = tokio::spawn(import_future);

select! {
    // ^C thing
    // progress thing
}

real_work_handle.await?;
```

The difference I see is inlining the work that's spread between the progress bar and ctrl_c watch functions which don't seem to be particularly self-contained.

Perhaps it's about the same, but I wanted to float the idea.

Reply:
That'a bit cleaner, I combined the two tasks.